### PR TITLE
Make UnitSpec generic parameter for `LayerSpec` and `FacetSpec`

### PR DESCRIPTION
**Pull request by kanitw**

### CODE REVIEW

File: src/spec.ts [hunk 1]
```diff
@@ -79,10 +79,12 @@ export interface GenericUnitSpec<M, E extends Encoding> extends BaseSpec {
 
 export type UnitSpec = GenericUnitSpec<Mark, Encoding>;
 
+export type LayeredUnitSpec = GenericUnitSpec<AnyMark, Encoding>;
+
 
 export type FacetedUnitSpec = GenericUnitSpec<AnyMark, EncodingWithFacet>;
 
-export interface GenericLayerSpec<M, E> extends BaseSpec {
+export interface GenericLayerSpec<U> extends BaseSpec {
```
Comment (domoritz) [inline src/spec.ts]: You can put constraints on generics. See https://www.typescriptlang.org/docs/handbook/generics.html

`GenericLayerSpec<U extends BaseSpec>`?

File: vega-lite-schema.json [hunk 2]
```diff
@@ -1341,7 +1341,7 @@
             ],
             "type": "object"
         },
-        "GenericFacetSpec<AnyMark, EncodingWithFacet>": {
+        "GenericFacetSpec<GenericUnitSpec<AnyMark, EncodingWithFacet>>": {
```
Comment (domoritz) [inline vega-lite-schema.json]: That's barely understandable.

Comment (kanitw) [inline vega-lite-schema.json]: I guess you can modify your generator to name things based on their alias.

File: src/spec.ts (hunk 1 above)
Comment (kanitw) [inline src/spec.ts]: I just make another commit for `U extends GenericUnitSpec`.  (Let's squash after we are happy with other things.)

File: vega-lite-schema.json (hunk 2 above)
Comment (domoritz) [inline vega-lite-schema.json]: Yeah, it's called --aliasRef :-P

Comment (kanitw) [inline vega-lite-schema.json]: Oh and this is something you say you should turn it on at some point?

Comment (domoritz) [inline vega-lite-schema.json]: Yes. So we get nice names :-P

Comment (domoritz) [inline vega-lite-schema.json]: But we need to make sure we don't break anything.

Comment (kanitw) [inline vega-lite-schema.json]: Then I guess we can merge this, assuming that we will turn that nice name flag on soon?
